index.js

```diff
@@ -82,46 +93,47 @@ function serveStatic(root, options) {
       return
     }
 
-    var forwardError = !fallthrough
     var originalUrl = parseUrl.original(req)
-    var path = parseUrl(req).pathname
+    var originalPath = parseUrl(req).pathname
 
     // make sure redirect occurs at mount
-    if (path === '/' && originalUrl.pathname.substr(-1) !== '/') {
-      path = ''
+    if (originalPath === '/' && originalUrl.pathname.substr(-1) !== '/') {
+      originalPath = ''
     }
 
-    // create send stream
-    var stream = send(req, path, opts)
-
-    // add directory handler
-    stream.on('directory', onDirectory)
+    // options passed to stream
+    var streamOptions = {
+      opts: opts,
+      fallthrough: fallthrough,
+      setHeaders: setHeaders,
+      onDirectory: onDirectory,
+      path: originalPath
+    }
 
-    // add headers listener
-    if (setHeaders) {
-      stream.on('headers', setHeaders)
+    // static gzip serving disabled
+    var fallbackStream = streamFile.bind(this, req, res, next, streamOptions)
+    if (!serveGzip) {
+      return fallbackStream()
     }
 
-    // add file listener for fallthrough
-    if (fallthrough) {
-      stream.on('file', function onFile() {
-        // once file is determined, always forward error
-        forwardError = true
-      })
+    // gzip encoding not supported by client
+    var acceptEncoding = req.headers['accept-encoding'] || ''
+    if (acceptEncoding.indexOf('gzip') === -1) {
+      return fallbackStream()
     }
 
-    // forward errors
-    stream.on('error', function error(err) {
-      if (forwardError || !(err.statusCode < 500)) {
-        next(err)
-        return
-      }
+    // static gzip file not found
+    var gzipPath = originalPath + '.gz'
+    if (!gzipCache[path.join(root, gzipPath)]) {
+      return fallbackStream()
+    }
 
-      next()
-    })
+    // set gzip specific headers
+    setGzipHeaders(res, streamOptions.path)
```

Comment:
Can this not use the send stream's `headers` event to add these headers? The purpose of that event is so you are only adding the header is the file is ultimately served; this ends up adding the headers unconditionally, and on things like 404 from the file disappearing to even a 500 from opening the file handle, headers already set like `Content-Encoding` will make the resulting 404/500 unusable, as the browser will be expecting that to be in gzip format.